packages/react/src/components/Button/Button.tsx

```diff
@@ -84,120 +97,177 @@ export interface ButtonProps
-        {Box.create(!hasChildren && content, {
-          defaultProps: () => ({ as: 'span', styles: styles.content }),
-        })}
-        {!hasChildren && iconPosition === 'after' && this.renderIcon(variables, styles)}
-      </ElementType>
-    )
-  }
+    variables,
+    design,
+  } = props
+  const context: ProviderContextPrepared = React.useContext(ThemeContext)
+  const { setStart, setEnd } = useTelemetry(Button.displayName, context.telemetry)
+
+  setStart()
+
+  const hasChildren = childrenExist(children)
+
+  const getA11Props = useAccessibility(accessibility, {
+    debugName: Button.displayName,
+    mapPropsToBehavior: () => ({
+      as,
+      active,
+      disabled,
+      loading,
+    }),
+    actionHandlers: {
+      performClick: event => {
+        event.preventDefault()
+        handleClick(event)
+      },
+    },
+    rtl: context.rtl,
+  })
+  const { classes, styles: resolvedStyles } = useStyles(Button.displayName, {
+    className: Button.className,
+    mapPropsToStyles: () => ({
+      text,
+      primary,
+      disabled,
+      circular,
+      size,
+      loading,
+      inverted,
+      iconOnly,
+      fluid,
+      hasContent: !!content,
+    }),
+    mapPropsToInlineStyles: () => ({
+      className,
+      design,
+      styles,
+      variables,
+    }),
+    rtl: context.rtl,
+  })
 
-  renderIcon = (variables, styles) => {
-    const { icon, iconPosition, content } = this.props
+  const unhandledProps = getUnhandledProps(Button.handledProps, props)
+  const ElementType = getElementType(props)
 
+  const renderIcon = () => {
     return Icon.create(icon, {
-      defaultProps: () => ({
-        styles: styles.icon,
-        xSpacing: !content ? 'none' : iconPosition === 'after' ? 'before' : 'after',
-        variables: variables.icon,
-      }),
+      defaultProps: () =>
+        getA11Props('icon', {
+          styles: resolvedStyles.icon,
+          xSpacing: !content ? 'none' : iconPosition === 'after' ? 'before' : 'after',
+        }),
     })
   }
 
-  renderLoader = (variables, styles) => {
-    const { loader } = this.props
-
+  const renderLoader = () => {
     return Loader.create(loader || {}, {
-      defaultProps: () => ({
-        role: undefined,
-        styles: styles.loader,
-      }),
+      defaultProps: () =>
+        getA11Props('loader', {
+          role: undefined,
+          styles: resolvedStyles.loader,
+        }),
     })
   }
 
-  handleClick = (e: React.SyntheticEvent) => {
-    const { disabled } = this.props
-
+  const handleClick = (e: React.SyntheticEvent) => {
     if (disabled) {
       e.preventDefault()
       return
     }
 
-    _.invoke(this.props, 'onClick', e, this.props)
+    _.invoke(props, 'onClick', e, props)
   }
 
-  handleFocus = (e: React.SyntheticEvent) => {
-    _.invoke(this.props, 'onFocus', e, this.props)
+  const handleFocus = (e: React.SyntheticEvent) => {
+    _.invoke(props, 'onFocus', e, props)
   }
+
+  setEnd()
```

Comment:
This is extremely fragile and (currently) wrong.
What if I forget to add telemetry to my component? There should be a conformance test for that.

But the main problem I see is that nothing prevents you from calling `setStart` too late or `setEnd` too early.

In this case the whole return is down outside of the telemetry scope - all calls to `getAttributes()`, `getA11Props()`, `renderLoader()`, `renderIcon()`... - which makes the numbers useless :-(

So we should either give up and remove fluent-ui telemetry (no!) or rethink and rework how we measure it (something like a wrapping function for the whole render? And named something like `renderComponent`?)

Reply:
No `renderComponent` please, we are trying to move out from it. This is an internal thing and we can track it (like you can block the PR because it's not in the Avatar, and Image components :D - oops will ad it) Even when this was in renderComponent, it was setting end even before invoking render() which may be even worse.. Anyway, I will add the telemetry to the other components before merging. Do you have some other proposal for the setting of the start and end?

Reply:
> Even when this was in renderComponent, it was setting end even before invoking render() which may be even worse..

That's not true. That's what @layershifter broke in #1980

Reply:
Fixed renderComponent, fixed components' telemetry to be accurate. It is used internally now, so we can make sure we are adding it properly until we have better idea of how to improve it.